torch/optim/optimizer.py

```diff
@@ -393,6 +394,27 @@ def pack_group(group):
             'param_groups': param_groups,
         }
 
+    @staticmethod
+    def _process_value_according_to_param_policy(param: Tensor, value: Tensor, paramID: int = None,
+                                                 paramGroups: List[Dict[Any, Any]] = None, key=None) -> Tensor:
+        # Floating-point types are a bit special here. They are the only ones
+        # that are assumed to always match the type of params.
+        # Make sure state['step'] is not casted https://github.com/pytorch/pytorch/issues/74424
+        # UNLESS fused or capturable, see [special device hosting for step]
+        fused = False
+        capturable = False
+        for pg in paramGroups:
+            if paramID in pg["params"]:
+                fused = pg["fused"] if "fused" in pg else False
+                capturable = pg["capturable"] if "capturable" in pg else False
+                break
+
+        if key != "step" or capturable or fused:
+            if param.is_floating_point():
+                value = value.to(param.dtype)
+            value = value.to(param.device)
```

Comment:
nit but could be significant runtime
```suggestion
            if param.is_floating_point():
                value = value.to(dtype=param.dtype, device=param.device)
            else:
                value = value.to(device=param.device)
```

Reply:
oooh good call